extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,361 @@
+		this._nodeType = nodeType;
+	}
+
+	public set iconPath(iconPath: { dark: string, light: string }) {
+		this._iconPath = iconPath;
+	}
+
+	public get iconPath(): { dark: string, light: string } {
+		return this._iconPath;
+	}
+
+	public set treeChangeHandler(treeChangeHandler: IControllerTreeChangeHandler) {
+		this._treeChangeHandler = treeChangeHandler;
+	}
+
+	public get treeChangeHandler(): IControllerTreeChangeHandler {
+		return this._treeChangeHandler;
+	}
+}
+
+export class ControllerRootNode extends ControllerTreeNode {
+
+	constructor(properties?: { treeChangeHandler: IControllerTreeChangeHandler }) {
+		super(Object.assign({
+			label: 'root',
+			nodeType: BdcItemType.controllerRoot,
+		}, properties));
+	}
+
+	public async getChildren(): Promise<ControllerNode[]> {
+		return this.children as ControllerNode[];
+	}
+
+	public addControllerNode(url: string, username: string, password: string, rememberPassword: boolean, masterInstance?: IEndPoint): void {
+		let controllerNode = this.getExistingControllerNode(url, username);
+		if (controllerNode) {
+			controllerNode.password = password;
+			controllerNode.rememberPassword = rememberPassword;
+			controllerNode.clearChildren();
+		} else {
+			controllerNode = new ControllerNode({ url, username, password, rememberPassword, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(controllerNode);
+		}
+
+		if (masterInstance) {
+			controllerNode.addSqlMasterNode(masterInstance.endpoint, masterInstance.description);
+		}
+	}
+
+	public deleteControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		let index = nodes.findIndex(e => e.url === url && e.username === username);
+		let deleted = undefined;
+		if (index >= 0) {
+			deleted = nodes.splice(index, 1);
+		}
+		return deleted;
+	}
+
+	private getExistingControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		return nodes.find(e => e.url === url && e.username === username);
+	}
+}
+
+export class ControllerNode extends ControllerTreeNode {
+	private _url: string;
+	private _username: string;
+	private _password: string;
+	private _rememberPassword: boolean;
+
+	constructor(properties?: {
+		url: string,
+		username: string,
+		password: string,
+		parent: ControllerTreeNode,
+		treeChangeHandler: IControllerTreeChangeHandler,
+		label?: string,
+		description?: string,
+		rememberPassword?: boolean
+	}) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.controller,
+			iconPath: IconPath.controllerNode
+		}, properties));
+
+		let address = ControllerNode.toIpAndPort(properties.url);
+		this.label = properties.label || `controller: ${address} (${properties.username})`;
+		this.description = properties.description || this.label;
```

Comment:
Localize. Should controller also be capitalized?

Reply:
lower case was accepted in design review. They will let me know if they want to change.